### PR TITLE
API and UI for configuring number of digits to display on numbers

### LINKED ISSUE #1332: Use display options (instead of db options) to format decimal places
## Our current design

- When Postgres `scale` is **not** defined, the number is formatted with however many decimal places exist in the stored value.

- When Postgres `scale` **is** defined, the number is formatted to have exactly that number of decimal places, regardless of the stored value.

### (A) Format with fewer decimal places than are stored

Spreadsheets allow this, and in my experience it's quite common to utilize this feature.

Peek 2022-04-28 21-30

For example, let's say my data is:

| v |
| - |
| 123.4 |
| 9.809215179162 |

To work with this data easily, I'd like to see:

| v |
| -: |
| 123.40 |
| 9.81 |

(Notice that the second value is _rounded_ to `9.81` instead of being truncated to `9.80`)

But I don't want to lose the precision of that second value just to make my life easier looking at the data with all the decimal points lined up. 

I'd like to apply _formatting_ that makes it easy for me to view my data, without changing the actual data.

### (B) Format a floating point number

The Postgres `scale` property does not exist for `real` or `double precision` numbers. Users may want to format values of that type -- especially if we adopt my proposal in #1331.

### (C) Specify minimum digits independently of maximum digits

Let's pretend I'm storing money amounts in USD and I have the following values:

| dollars |
| - |
| 1.4 |
| 2.8191 |

Also, for my task at-hand I'd like to see the full values, with all their stored precision, but I don't happen to know the maximum number of decimal places in my data set (and indeed it could change as I add new data). For this reason I don't want to set a _maximum_ number of decimal places. But I _do_ want to set a _minimum_ number of decimal places to 2 because it's currency. I'd like to see:

| dollars |
| -: |
| $1.40 |
| $2.8191 |

## Proposal

- Add two display options `minimum_fraction_digits` and `maximum_fraction_digits`, which correspond to the the equivalently named options in the browser [Intl API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat)

- Design a UI for the user to configure their decimal settings through these display options (instead of through the `scale` and `precision` options).

- To keep the UI simple, we could begin with one input that sets both `minimum_fraction_digits` and `maximum_fraction_digits` in tandem, thereby satisfying (A) and (B), but not (C). Then later (with more work), we could add a separate checkbox that turns the single input into two separate inputs for more advanced use-cases (without needing to change the display_options schema). (I can clarify this with a mockup if it's confusing.)

- Continue to expose `scale` and `precision`, but hide them behind a checkbox and supply help text to explain their purpose while dissuaded users from setting values there.

### PR DESCRIPTION
**Pull request by seancolsen**
Fixes #1332

## Before

- The number of digits displayed for a number is always taken from the number of digits in that number's numerical value. There is no way to format `1` as `1.00`, for example. And no way to format `1.119999999` as `1.12`.

## After

- The back end and API supports two display options for numbers:

    - `minimum_fraction_digits`
    - `maximum_fraction_digits`

    Why two separate options? See use-case "(C)" in the [ticket](https://github.com/centerofci/mathesar/issues/1332).

- The Postgres `scale` value for `NUMERIC` numbers is _not_ used to format the number of decimal places. A benefit to this approach is that we can format other Postgres types like `DOUBLE PRECISION`.

- If the value of a number has more decimal places than the formatting allows, it will be rounded during formatting, and the full value will be visible while editing.

## Demo

https://www.loom.com/share/ddd468c2c40e42d190632e56676e8827

## Notes

- The UI implemented here does not address use-case "(C)" in the [ticket](https://github.com/centerofci/mathesar/issues/1332). I have opened #1401 for that purpose.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by seancolsen** [comment]
This is a full-stack PR, so I've assigned @silentninja to review the back end (because this feature is in the data types milestone), and @pavish to review the front end.

**Review by silentninja — CHANGES_REQUESTED** [review]
Looks good to me. Got a few minor changes which I have commented on below.

File: mathesar/api/serializers/shared_serializers.py [hunk 1]
```diff
@@ -134,6 +147,24 @@ class AbstractNumberDisplayOptionSerializer(serializers.Serializer):
...
+    maximum_fraction_digits = serializers.IntegerField(**FRACTION_DIGITS_CONFIG)
+
+    def _validate_fraction_digits(self, data):
+        minimum = data.get("minimum_fraction_digits")
+        maximum = data.get("maximum_fraction_digits")
+        if minimum is None or maximum is None:
+            # No errors if one of the fields is not set
+            return
+        if minimum > maximum:
+            raise serializers.ValidationError(
```
[the remark this answers is not in the document]
Comment (seancolsen) [reply mathesar/api/serializers/shared_serializers.py]: Right. Thanks! This should be fixed in 3c978af758ec50674015e52cf0df328f03835543

**Comment by seancolsen** [comment]
@silentninja this is ready for re-review.

**Review by pavish — APPROVED** [review]
@seancolsen Looks great to me!

I think having `Decimal places` in both type options and display options would be quite confusing to the user. I'd lean towards removing `Decimal places` and `Max digits` type options entirely on the frontend (scale and precision).

**Comment by seancolsen** [comment]
@pavish 

> I think having Decimal places in both type options and display options would be quite confusing to the user.

Agreed. I've opened #1420 for this.

**Pull request merged**